buildpack/files.go

```diff
@@ -198,3 +202,111 @@ type LayerMetadataFile struct {
 	Launch bool        `json:"launch" toml:"launch"`
 	Cache  bool        `json:"cache" toml:"cache"`
 }
+
+type typesTable struct {
+	Build  bool `toml:"build"`
+	Launch bool `toml:"launch"`
+	Cache  bool `toml:"cache"`
+}
+type layerMetadataTomlFile struct {
+	Data  interface{} `toml:"metadata"`
+	Types typesTable  `toml:"types"`
+}
+
+func (lmf *LayerMetadataFile) Encode(path, buildpackAPI string) error {
+	fh, err := os.Create(path)
+	if err != nil {
+		return err
+	}
+	defer fh.Close()
+	if supportsTypesTable(buildpackAPI) {
+		types := typesTable{Build: lmf.Build, Launch: lmf.Launch, Cache: lmf.Cache}
+		lmtf := layerMetadataTomlFile{Data: lmf.Data, Types: types}
+		return toml.NewEncoder(fh).Encode(lmtf)
+	}
+	return toml.NewEncoder(fh).Encode(lmf)
+}
+
+func DecodeLayerMetadataFile(path, buildpackAPI string) (LayerMetadataFile, error) {
+	fh, err := os.Open(path)
+	if os.IsNotExist(err) {
+		return LayerMetadataFile{}, nil
+	} else if err != nil {
+		return LayerMetadataFile{}, err
+	}
+	defer fh.Close()
+
+	if supportsTypesTable(buildpackAPI) {
+		var lmtf layerMetadataTomlFile
+		if _, err = toml.DecodeFile(path, &lmtf); err != nil {
+			return LayerMetadataFile{}, err
+		}
+		return LayerMetadataFile{Data: lmtf.Data, Build: lmtf.Types.Build, Launch: lmtf.Types.Launch, Cache: lmtf.Types.Cache}, nil
+	}
+	var lmf LayerMetadataFile
+	if _, err := toml.DecodeFile(path, &lmf); err != nil {
```

Comment:
Following your idea [here](https://github.com/buildpacks/lifecycle/pull/537#discussion_r585936785), there are 3 `toml.DecodeFile` in this file.
2 in the `DecodeLayerMetadataFile` function, each one to decode to the right format.
1 in the `typesInRightFormat` function including checking the return value of `toml.MetaData`.
Do you think there is a way to reduce it while keeping the code clear and simple?

Reply:
What do you think about the following idea:
Having one function `DecodeLayerMetadataFile` that returning `LayerMetadataFile, error, bool`.
The boolean will be `true` if the toml file was in the expected format depending on the `buildpack API`, and `false` otherwise.
Also, when calling `toml.DecodeFile`, it'll be into the following `struct`:
type layerMetadataTomlFile struct {
	Data  interface{} `toml:"metadata"`
	Types typesTable  `toml:"types"`
         Build  bool        `json:"build" toml:"build"`
	Launch bool        `json:"launch" toml:"launch"`
	Cache  bool        `json:"cache" toml:"cache"`
}
In this way we won't lose data and we'll be able to call `toml.DecodeFile` only once.

Reply:
I was just thinking about a small utility function to at least reduce the boilerplate around `toml.Decode` and the pointer
```
func decodeLayerTOMLFile(path string) (layerMetadataTomlFile, error)
{
    var lmtf layerMetadataTomlFile
    if _, err = toml.DecodeFile(path, &lmtf); err != nil {
        return LayerMetadataFile{}, err
    }
    return lmtf, nil
}
```

Calling code is just a bit shorter
```
if supportsTypesTable(buildpackAPI) {
     lmtf, err := decodeLayerTOMLFile(path); 
     if err != nil {
	return LayerMetadataFile{}, err
     }
     return LayerMetadataFile{Data: lmtf.Data, Build: lmtf.Types.Build, Launch: lmtf.Types.Launch, Cache: lmtf.Types.Cache}, nil

}
return decodeLayerTOMLFile(path);
```
It may not be worth it, but that is what I had in mind when I wrote the comment!

Reply:
mmmmm but I think that `decodeLayerTOMLFile` cannot be called as is for old buildpack API and for new buildpack API.
Right?

Reply:
Your idea seems like a good path to explore! 

We could also explore two distinct structs and them both implementing a `LayerTOML` interface `type LayerTOMLFile interface { Build() bool, Launch() bool, Warnings() []string etc...}`. A single parse driven by the Buildpack API and the translation to the interface would be implemented by the struct itself.

The metadata from the parsing of the toml could be stored in the structs and interrogated by a `Warnings() []string` method when needed. If you want to explore something like that, I can try and pair with you sometime. It might be bigger than we are ready to tackle for this PR tho 😛 

```
func (l layerTOMLv2) Build() bool {
 return l.Types.Build
}

func (l layerTOMLv1) Warnings() []string {
 if l.parseMetadata.IsDefined("types") { ...  }
}
```

Reply:
Interesting!
I saw your comment after pushing the commit https://github.com/buildpacks/lifecycle/pull/537/commits/de2f8e688570dc8f2c17f94b92b944c5652cc6e0.
Please review it and let's discuss your idea vs my commit tomorrow.

Reply:
Nice improvements! Another struct makes the intent clear and the mapping is explicit.